backend/manager/modules/restapi/jaxrs/src/main/java/org/ovirt/engine/api/restapi/resource/BackendHostCpuUnitsResource.java

```diff
@@ -12,4 +27,94 @@ public BackendHostCpuUnitsResource(Guid hostId) {
+
+    @Override
+    public HostCpuUnits list() {
+        VDS host = getHost();
+
+        if (host.getCpuTopology() == null) {
+            return new HostCpuUnits();
+        }
+
+        List<VM> vms = getMonitoredVms();
+
+        HostCpuUnits hostCpuUnits = new HostCpuUnits();
+        Mapper<VdsCpuUnit, HostCpuUnit> mapper = getMapper(VdsCpuUnit.class, HostCpuUnit.class);
+        for (VdsCpuUnit vdsCpuUnit : host.getCpuTopology()) {
+            hostCpuUnits.getHostCpuUnits().add(mapper.map(vdsCpuUnit, null));
+        }
+
+        addPinnedVms(hostCpuUnits.getHostCpuUnits(), vms);
+        addVdsAffinity(hostCpuUnits.getHostCpuUnits(), host.getVdsmCpusAffinity());
+
+        hostCpuUnits.getHostCpuUnits().forEach(this::addLinks);
+        return hostCpuUnits;
+    }
+
+    private void addPinnedVms(List<HostCpuUnit> hostCpuUnits, List<VM> vms) {
+        for (VM vm : vms) {
+            String vmPinning = CpuPinningHelper.getVmPinning(vm);
+            if (vmPinning == null) {
+                continue;
+            }
+            Set<Integer> vmPinnedCpus = CpuPinningHelper.getAllPinnedPCpus(vmPinning);
+            hostCpuUnits.stream()
+                    .filter(hostCpuUnit -> vmPinnedCpus.contains(hostCpuUnit.getCpuId()))
+                    .forEach(hostCpuUnit -> {
+                        Vm v = new Vm();
+                        v.setId(vm.getId().toString());
+                        v.setName(vm.getName());
+                        v.setCpuPinningPolicy(VmMapper.map(vm.getCpuPinningPolicy()));
+
+                        if (hostCpuUnit.getVms() == null) {
+                            hostCpuUnit.setVms(new Vms());
+                        }
+                        hostCpuUnit.getVms().getVms().add(v);
+                    });
+        }
+    }
+
+    private void addVdsAffinity(List<HostCpuUnit> hostCpuUnits, String affinityString) {
+        Integer affinity = parseAffinity(affinityString);
+        hostCpuUnits.stream()
+                .filter(hostCpuUnit -> hostCpuUnit.getCpuId() == affinity)
+                .forEach(hostCpuUnit -> hostCpuUnit.setRunsVdsm(true));
+    }
+
+    private int parseAffinity(String affinity) {
+        try {
+            return Integer.parseInt(affinity);
+        } catch (Exception e) {
+            return -1;
+        }
+    }
+
+    private VDS getHost() {
+        QueryReturnValue queryReturnValue = runQuery(QueryType.GetVdsByVdsId, new IdQueryParameters(hostId));
+        if (!queryReturnValue.getSucceeded()) {
+            try {
+                backendFailure(queryReturnValue.getExceptionString());
+            } catch (BackendFailureException ex) {
+                handleError(ex, false);
+            }
+        }
+
+        VDS host = queryReturnValue.getReturnValue();
+        return host;
+    }
+
+    private List<VM> getMonitoredVms() {
+        QueryReturnValue queryReturnValue =
+                runQuery(QueryType.GetMonitoredVmsRunningOnVds, new IdQueryParameters(hostId));
```

Comment:
why not to call `GetAllVmsRunningForMultipleVdsQuery`?
the monitored vms were meant for monitoring, where we're ok with getting partial information in order to minimize the load on the database - but this one is not expected to be called frequently